sdk/core/azure-core-tracing-opentelemetry/azure/core/tracing/ext/opentelemetry_span/__init__.py

```diff
@@ -39,37 +39,50 @@
 __version__ = VERSION
 
 
-def _get_headers_from_http_request_headers(headers: "Mapping[str, Any]", key: str):
-    """Return headers that matches this key.
-
-    Must comply to opentelemetry.context.propagation.httptextformat.Getter:
-    Getter = typing.Callable[[_T, str], typing.List[str]]
-    """
-    return [headers.get(key, "")]
-
-
-def _set_headers_from_http_request_headers(headers: "MutableMapping[str, Any]", key: str, value: str):
-    """Set headers in the given headers dict.
-
-    Must comply to opentelemetry.context.propagation.httptextformat.Setter:
-    Setter = typing.Callable[[_T, str, str], None]
-    """
-    headers[key] = value
-
-
 class OpenTelemetrySpan(HttpSpanMixin, object):
     """OpenTelemetry plugin for Azure client libraries.
 
     :param span: The OpenTelemetry span to wrap, or nothing to create a new one.
     :type span: ~OpenTelemetry.trace.Span
     :param name: The name of the OpenTelemetry span to create if a new span is needed
     :type name: str
+    :keyword SpanKind kind: The span kind of this span.
+    :keyword links: The list of links to be added to the span.
+    :paramtype links: list[~azure.core.tracing.Link]
     """
 
-    def __init__(self, span=None, name="span"):
-        # type: (Optional[Span], Optional[str]) -> None
+    def __init__(self, span=None, name="span", **kwargs):
+        # type: (Optional[Span], Optional[str], Any) -> None
         current_tracer = self.get_current_tracer()
-        self._span_instance = span or current_tracer.start_span(name=name)
+
+        ## kind
+        value = kwargs.pop('kind', None)
+        kind = (
+            OpenTelemetrySpanKind.CLIENT if value == SpanKind.CLIENT else
+            OpenTelemetrySpanKind.PRODUCER if value == SpanKind.PRODUCER else
+            OpenTelemetrySpanKind.SERVER if value == SpanKind.SERVER else
+            OpenTelemetrySpanKind.CONSUMER if value == SpanKind.CONSUMER else
+            OpenTelemetrySpanKind.INTERNAL if value == SpanKind.INTERNAL else
+            OpenTelemetrySpanKind.INTERNAL if value == SpanKind.UNSPECIFIED else
+            None
+        ) # type: SpanKind
+        if value and kind is None:
+            raise ValueError("Kind {} is not supported in OpenTelemetry".format(value))
+
+        links = kwargs.pop('links', None)
+        if links:
+            try:
+                ot_links = []
+                for link in links:
+                    ctx = extract(link.headers)
+                    span_ctx = get_span_from_context(ctx).get_span_context()
+                    ot_links.append(OpenTelemetryLink(span_ctx, link.attributes))
+                    kwargs.setdefault('links', ot_links)
```

Comment:
Why is this done inside the loop? Looks like it should be outside, no?

Reply:
this should be outside - moved it (y)